File: content/problems/3098-find-the-sum-of-subsequence-powers/solutions/threshold-counting-dp-over-distinct-differences/Solution.java
class Solution {
    private static final long MOD=1000000007L;
    private long countAtLeast(int[] nums,int k,int difference){
        int n=nums.length;
        long[] previous=new long[n];
        Arrays.fill(previous,1);
        for(int length=2;length<=k;length++){
            long[] prefix=new long[n],current=new long[n];
            long sum=0;
            for(int index=0;index<n;index++){sum=(sum+previous[index])%MOD;prefix[index]=sum;}
            int pointer=-1;
            for(int index=0;index<n;index++){
                while(pointer+1<index&&(long)nums[index]-nums[pointer+1]>=difference)pointer++;
                if(pointer>=0)current[index]=prefix[pointer];
            }
            previous=current;
        }
        long result=0;
        for(long count:previous)result=(result+count)%MOD;
        return result;
    }
    public int sumOfPowers(int[] nums,int k){
        Arrays.sort(nums);
        TreeSet<Integer> set=new TreeSet<>();
        for(int left=0;left<nums.length;left++)for(int right=left+1;right<nums.length;right++)if(nums[right]>nums[left])set.add(nums[right]-nums[left]);
        int[] differences=new int[set.size()];
        int write=0;
        for(int difference:set)differences[write++]=difference;
        long[] counts=new long[differences.length];
        for(int index=0;index<differences.length;index++)counts[index]=countAtLeast(nums,k,differences[index]);
        long answer=0;
        for(int index=0;index<differences.length;index++){
            long next=index+1<differences.length?counts[index+1]:0;
            long exact=(counts[index]-next+MOD)%MOD;
            answer=(answer+(differences[index]%MOD)*exact)%MOD;
        }
        return (int)answer;
    }
}

File: content/problems/3098-find-the-sum-of-subsequence-powers/solutions/threshold-counting-dp-over-distinct-differences/Solution.kt
class Solution {
    private val mod=1000000007L
    private fun countAtLeast(nums:IntArray,k:Int,difference:Int):Long{
        val n=nums.size
        var previous=LongArray(n){1L}
        for(length in 2..k){
            val prefix=LongArray(n)
            val current=LongArray(n)
            var sum=0L
            for(index in 0 until n){sum=(sum+previous[index])%mod;prefix[index]=sum}
            var pointer=-1
            for(index in 0 until n){
                while(pointer+1<index&&nums[index].toLong()-nums[pointer+1]>=difference)pointer++
                if(pointer>=0)current[index]=prefix[pointer]
            }
            previous=current
        }
        var result=0L
        for(count in previous)result=(result+count)%mod
        return result
    }
    fun sumOfPowers(nums:IntArray,k:Int):Int{
        nums.sort()
        val set=java.util.TreeSet<Int>()
        for(left in nums.indices)for(right in left+1 until nums.size)if(nums[right]>nums[left])set.add(nums[right]-nums[left])
        val differences=set.toList()
        val counts=LongArray(differences.size){countAtLeast(nums,k,differences[it])}
        var answer=0L
        for(index in differences.indices){
            val next=if(index+1<counts.size)counts[index+1] else 0L
            val exact=(counts[index]-next+mod)%mod
            answer=(answer+differences[index].toLong()%mod*exact)%mod
        }
        return answer.toInt()
    }
}
